### PR TITLE
relax trl 0.x upper bound from <=0.24.0 to <0.30.0

### PR DESCRIPTION
**Pull request by shylane**
# Relax `trl` 0.x upper bound from `<=0.24.0` to `<0.30.0`

## Problem

`unsloth-zoo==2026.5.1` and current `main` pin:

```toml
trl>=0.18.2,!=0.19.0,<=0.24.0
```

This blocks users who intentionally need TRL 0.25-0.29, even though these versions remain in the pre-1.0 line. For example, post-0.24 GRPO work includes newer configuration and trainer surface such as `multi_objective_aggregation`, SAPO-related loss knobs, `off_policy_mask_threshold`, vLLM configuration updates, and optional tool/environment rollout support.

With the current cap, dependency resolution fails:

```bash
uv pip install "unsloth-zoo==2026.5.1" "trl==0.29.1"
# -> No solution found: unsloth-zoo requires trl<=0.24.0
```

## Compatibility notes

The `<=0.24.0` ceiling appears stricter than the current package metadata needs:

- `GRPOTrainer.__init__` in `trl==0.29.1` keeps the existing required arguments (`model`, `reward_funcs`) and only adds optional parameters (`tools`, `rollout_func`, `environment_factory`).
- `GRPOConfig` in both `trl==0.24.0` and `trl==0.29.1` exposes `num_iterations` and `importance_sampling_level`.
- `num_mini_batches` is not used by `unsloth-zoo`.
- `unsloth-zoo` already runtime-detects newer TRL GRPO loss features (`get_sapo_token_loss`, `get_off_policy_mask`, `sapo_temperature_*`, `off_policy_mask_threshold`) via `hasattr` guards in `rl_replacements.py`.

This PR intentionally keeps an upper bound at `<0.30.0`. TRL 1.x is a larger version jump and should be evaluated separately.

## Fix

Change `<=0.24.0` to `<0.30.0` in both `pyproject.toml` dependency lists:

```diff
-    "trl>=0.18.2,!=0.19.0,<=0.24.0 ; (sys_platform != 'darwin' or platform_machine != 'arm64')",
+    "trl>=0.18.2,!=0.19.0,<0.30.0 ; (sys_platform != 'darwin' or platform_machine != 'arm64')",
```

and in `[project.optional-dependencies] core`:

```diff
-    "trl>=0.18.2,!=0.19.0,<=0.24.0",
+    "trl>=0.18.2,!=0.19.0,<0.30.0",
```

## Testing

Environment: Python 3.11.14, uv 0.9.16.

Current published metadata fails as expected:

```bash
uv pip install --dry-run "unsloth-zoo==2026.5.1" "trl==0.29.1"
# -> No solution found: unsloth-zoo requires trl<=0.24.0
```

Patched local checkout resolves:

```bash
uv pip install --dry-run "./unsloth-zoo" "trl==0.29.1"
# -> Resolved successfully
```

Also checked:

| `trl` version | Result with patched constraint |
|---------------|--------------------------------|
| `0.24.0`      | resolves                       |
| `0.27.2`      | resolves                       |
| `0.29.1`      | resolves                       |
| `1.4.0`       | rejected by `<0.30.0`          |

The patched checkout was also resolver-tested with `--python-platform linux` for `trl==0.29.1`.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -28,7 +28,7 @@ dependencies = [
     "triton>=3.0.0 ; ('linux' in sys_platform)",
     "tyro ; (sys_platform != 'darwin' or platform_machine != 'arm64')",
     "accelerate>=0.34.1 ; (sys_platform != 'darwin' or platform_machine != 'arm64')",
-    "trl>=0.18.2,!=0.19.0,<=0.24.0 ; (sys_platform != 'darwin' or platform_machine != 'arm64')",
+    "trl>=0.18.2,!=0.19.0,<=0.30.0 ; (sys_platform != 'darwin' or platform_machine != 'arm64')",
```
[the remark this answers is not in the document]
Comment (shylane) [reply pyproject.toml]: Fixed in the latest commit — the specifier is now correctly <0.30.0.

[the remark this answers is not in the document]
Comment (shylane) [reply pyproject.toml]: Fixed in the latest commit — the specifier is now correctly <0.30.0.

File: pyproject.toml [hunk 2]
```diff
@@ -80,7 +80,7 @@ core = [
     "wheel>=0.42.0",
     "numpy",
     "accelerate>=0.34.1",
-    "trl>=0.18.2,!=0.19.0,<=0.24.0",
+    "trl>=0.18.2,!=0.19.0,<=0.30.0",
```
[the remark this answers is not in the document]
Comment (shylane) [reply pyproject.toml]: Fixed in the latest commit — the specifier is now correctly <0.30.0.

**Comment by danielhanchen** [comment]
Confirmed pyproject.toml still caps trl at <=0.24.0 in both the main and core dependency lists, and the newer GRPO knobs are already hasattr-guarded in rl_replacements.py. Could you note which trl minors between 0.25 and 0.29 you actually ran a GRPO training step on, not just resolved?

**Comment by shylane** [comment]
> Confirmed pyproject.toml still caps trl at <=0.24.0 in both the main and core dependency lists, and the newer GRPO knobs are already hasattr-guarded in rl_replacements.py. Could you note which trl minors between 0.25 and 0.29 you actually ran a GRPO training step on, not just resolved?


<p>Fair ask — the PR body's testing table was dependency-resolution only (<code>uv pip install --dry-run</code>), not a functional run. Just fixed that gap: ran an actual GRPO training step locally — Qwen2.5-1.5B-4bit + LoRA on an RTX 4060, <code>GRPOTrainer.train()</code> for 10 steps under bf16 autocast with a dummy length-based reward — against two points in the newly-allowed range:</p>

trl | Result | train_loss | last step loss
-- | -- | -- | --
0.27.2 | PASS (10/10 steps) | -0.03339 | -0.0337
0.29.1 | PASS (10/10 steps) | -0.03339 | -0.0337


<p>Identical loss trajectory at both versions, no crashes, and <code>rl_replacements.py</code>'s <code>hasattr</code> guards resolved cleanly on both. (Side note, unrelated to this PR: <code>trl==0.27.2</code> itself has an undeclared hard <code>import weave</code> in <code>callbacks.py</code> — had to <code>pip install weave</code> separately to even import <code>GRPOTrainer</code>. Not something unsloth-zoo controls, but worth knowing if others test this range.)</p>
<p>Happy to share the test script or run 0.25.x too if useful before merging.</p>

**Comment by danielhanchen** [comment]
Confirmed pyproject.toml still caps trl at <=0.24.0, and this would let new installs resolve 0.25-0.29 rather than only users who pin it themselves; could you share a training run (SFT and GRPO) on 0.29.1 before I review?

**Comment by danielhanchen** [comment]
Confirmed pyproject.toml still caps trl at 0.24.0, so pinning 0.25 to 0.29 fails to resolve; the cap is deliberate though, so could you say which TRL versions in that range you have actually run GRPO and SFT training on end to end?

**Comment by shylane** [comment]
You’re right to ask. I reran actual trainer tests, but the results do not establish stable end-to-end support across the entire `0.25–0.29` range.

Using this PR checkout, `unsloth==2026.5.1`, Qwen2.5-0.5B-Instruct with LoRA, Python 3.11, and an RTX 4060 Laptop GPU:

- I ran three-step SFT smoke tests on TRL `0.25.1`, `0.26.2`, `0.27.2`, `0.28.0`, and `0.29.1`, with `0.24.0` as a control. Every trainer run completed, although the short FP16 runs had some non-finite gradient-norm readings, including on the `0.24.0` control.
- I ran one-step GRPO smoke tests, covering generation, reward computation, backward propagation, and an optimizer step, on the same versions. The runs completed with finite gradients on `0.25.1`, `0.26.2`, and `0.29.1`.
- On `0.27.2` and `0.28.0`, the base installations failed while importing `GRPOTrainer` because `weave` was missing. Both one-step runs completed after installing `weave`.
- Separately, on `0.29.1`, a ten-step BF16 SFT run completed with finite losses and gradient norms throughout. A ten-step FP16 GRPO run completed with finite losses, but two of ten logged gradient norms were `NaN`, so I would not call that GRPO result numerically clean. These longer runs used the full model without 4-bit loading.

I also confirmed that this Zoo-only change does not make a normal installation resolve: the main `unsloth` package still requires `trl<=0.24.0`. I overrode that constraint only for these runtime tests.

These results provide concrete but limited smoke-test coverage for the tested releases; they do not establish stable training across every configuration permitted by `<0.30.0`.

This PR still has standalone relevance to direct `unsloth-zoo` consumers, but ordinary Unsloth installations would also require the corresponding constraint change in the main `unsloth` package. Would you prefer that this proceed as a coordinated Zoo plus `unsloth` change, with the exact test matrix and the `0.27.2`/`0.28.0` `weave` limitation documented, or should the current cap remain until broader CI coverage is available?
